src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -381,115 +353,131 @@ public override int Read(byte[] array, int offset, int count)
+            _writeQueue.Enqueue(result);
+            return result.Task;
         }
 
-        internal unsafe void Write(byte[] array, int offset, int count, int timeout)
+        public override IAsyncResult BeginRead(byte[] array, int offset, int numBytes, AsyncCallback userCallback, object stateObject)
         {
-            if (_inBreak)
-                throw new InvalidOperationException(SR.In_Break_State);
-            if (array == null)
-                throw new ArgumentNullException(nameof(array));
-            if (offset < 0)
-                throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedPosNum);
-            if (count < 0)
-                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedPosNum);
-            if (count == 0) return; // no need to expend overhead in creating asyncResult, etc.
-            if (array.Length - offset < count)
-                throw new ArgumentException(SR.Argument_InvalidOffLen);
+            return TaskToApm.Begin(ReadAsync(array, offset, numBytes), userCallback, stateObject);
+        }
 
-            // check for open handle, though the port is always supposed to be open
-            if (_handle == null) throw new ObjectDisposedException(SR.Port_not_open);
+        // Will wait `timeout` miliseconds or until reading or writing is possible
+        // If no operation is requested it will wait.
+        // Returns event which has happened
+        private Interop.Sys.PollEvents PollEvents(int timeout, bool pollReadEvents, bool pollWriteEvents)
+        {
+            if (!pollReadEvents && !pollWriteEvents)
+            {
+                Thread.Sleep(timeout);
+                return Interop.Sys.PollEvents.POLLNONE;
+            }
 
-            int numBytes = 0;
+            Interop.Sys.PollEvents eventsToPoll = Interop.Sys.PollEvents.POLLERR;
 
-            while (count > 0)
+            if (pollReadEvents)
             {
-                if (timeout > 0)
-                {
-                    Interop.Sys.PollEvents events = Interop.Sys.PollEvents.POLLNONE;
-                    Interop.Sys.Poll(_handle, Interop.Sys.PollEvents.POLLOUT | Interop.Sys.PollEvents.POLLERR, timeout,out events);
+                eventsToPoll |= Interop.Sys.PollEvents.POLLIN;
+            }
 
-                    if ((events & (Interop.Sys.PollEvents.POLLERR | Interop.Sys.PollEvents.POLLNVAL)) != 0)
-                    {
-                        throw new IOException();
-                    }
+            if (pollWriteEvents)
+            {
+                eventsToPoll |= Interop.Sys.PollEvents.POLLOUT;
+            }
 
-                    if ( (events & Interop.Sys.PollEvents.POLLOUT) == 0)
-                    {
-                        throw new TimeoutException(SR.Write_timed_out);
-                    }
-                }
+            Interop.Sys.PollEvents events = Interop.Sys.PollEvents.POLLNONE;
+            Interop.Sys.Poll(_handle,
```

Comment:
What if `poll` itself fails with an error?

Reply:
I'm slightly confused when could this happen - poll also returns POLLERR in revents when error happens and that is clear that it is related with fd but it is not clear to me when the poll itself can fail. Should we kill the loop at such case or perhaps retry and then kill if it keeps repeating or just always ignore. There is https://linux.die.net/man/3/explain_poll but it doesn't tell what kind of errors to expect or how to handle

Reply:
explained offline